django_ibmi/introspection.py

```diff
@@ -35,113 +30,71 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
-        indexes = {}
-        # To skip indexes across multiple fields
-        multifield_indexSet = set()
-        schema = cursor.connection.get_current_schema()
-        all_indexes = cursor.connection.indexes(True, schema, table_name)
-        for index in all_indexes:
-            if (index['ORDINAL_POSITION'] is not None) and (index['ORDINAL_POSITION'] == 2):
-                multifield_indexSet.add(index['INDEX_NAME'])
-
-        for index in all_indexes:
-            temp = {}
-            if index['INDEX_NAME'] in multifield_indexSet:
-                continue
-
-            if index['NON_UNIQUE']:
-                temp['unique'] = False
-            else:
-                temp['unique'] = True
-            temp['primary_key'] = False
-            indexes[index['COLUMN_NAME'].lower()] = temp
-
-        for index in cursor.connection.primary_keys(True, schema, table_name):
-            indexes[index['COLUMN_NAME'].lower()]['primary_key'] = True
-
-        return indexes
-
     # Getting the description of the table.
     def get_table_description(self, cursor, table_name):
         qn = self.connection.ops.quote_name
-        description = []
-        dbms_name = 'dbms_name'
-        schema = cursor.connection.get_current_schema()
-
-        if getattr(cursor.connection, dbms_name) == 'AS':
-            sql = "SELECT TYPE FROM QSYS2.SYSTABLES WHERE TABLE_SCHEMA='%(schema)s' AND TABLE_NAME='%(table)s'" % {
-                'schema': schema.upper(), 'table': table_name.upper()}
-        elif getattr(cursor.connection, dbms_name) != 'DB2':
-            sql = "SELECT TYPE FROM SYSCAT.TABLES WHERE TABSCHEMA='%(schema)s' AND TABNAME='%(table)s'" % {
-                'schema': schema.upper(), 'table': table_name.upper()}
-        else:
-            sql = "SELECT TYPE FROM SYSIBM.SYSTABLES WHERE CREATOR='%(schema)s' AND NAME='%(table)s'" % {
-                'schema': schema.upper(), 'table': table_name.upper()}
+        schema = self.connection.get_current_schema()
+        sql = "SELECT TYPE FROM QSYS2.SYSTABLES WHERE TABLE_SCHEMA='%s' AND TABLE_NAME='%s'" % \
+              (schema.upper(), table_name.upper())
         cursor.execute(sql)
         table_type = cursor.fetchone()[0]
 
+        field_map = {str(line.column_name): [line.data_type, str(line.column_def).replace("'", "")]
+                     for line in cursor.columns(schema=schema.upper(), table=table_name.upper())}
+
         if table_type != 'X':
             cursor.execute(
                 "SELECT * FROM %s FETCH FIRST 1 ROWS ONLY" % qn(table_name))
-            for desc in cursor.description:
-                description.append([desc[0].lower(), ] + desc[1:])
-        return description
+        return [
+            FieldInfo(
+                line[0],
+                field_map[str(line[0])][0],
+                *line[2:],
+                field_map[str(line[0])][1]
+            )
+            for line in cursor.description
+        ]
```

Comment:
This seems a bit more complicated than it needs to be. Looking at the sqlite3 backend, there's a description of what the API does:https://github.com/django/django/blob/76c0b32f826469320c59709d31e2f2126dd7c505/django/db/backends/sqlite3/introspection.py#L80-L84

So it's supposed to return what a PEP-249 cursor description does, which is defined here: https://www.python.org/dev/peps/pep-0249/#description

This information can be retrieved from [QSYS2.SYSCOLUMNS2](https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_71/db2/rbafzcatsyscol2.htm).

Reply:
Alternatively, call `cursor.columns`: https://github.com/mkleehammer/pyodbc/wiki/Features-beyond-the-DB-API#miscellaneous-odbc-functions